include/deal.II/matrix_free/cuda_fe_evaluation.cuh

```diff
@@ -0,0 +1,385 @@
+  evaluate(const bool evaluate_val, const bool evaluate_grad)
+  {
+    // First evaluate the gradients because it requires values that will be
+    // changed if evaluate_val is true
+    internal::EvaluatorTensorProduct<internal::EvaluatorVariant::evaluate_general,
+             dim, fe_degree,n_q_points_1d, Number> evaluator_tensor_product;
+    if (evaluate_grad == true)
+      {
+        evaluator_tensor_product.gradient_at_quad_pts(values, gradients);
+        __syncthreads();
+      }
+
+    if (evaluate_val == true)
+      {
+        evaluator_tensor_product.value_at_quad_pts(values);
+        __syncthreads();
+      }
+  }
+
+
+
+  template <int dim, int fe_degree, int n_q_points_1d, int n_components_,
+            typename Number>
+  __device__ void
+  FEEvaluation<dim, fe_degree, n_q_points_1d, n_components_, Number>::
+  integrate(const bool integrate_val, const bool integrate_grad)
+  {
+    internal::EvaluatorTensorProduct<internal::EvaluatorVariant::evaluate_general,
+             dim, fe_degree,n_q_points_1d, Number> evaluator_tensor_product;
+    if (integrate_val == true)
+      {
+        evaluator_tensor_product.integrate_value(values);
+        __syncthreads();
+        if (integrate_grad == true)
+          {
+            evaluator_tensor_product.integrate_gradient<true>(values, gradients);
+            __syncthreads();
+          }
+      }
+    else if (integrate_grad == true)
+      {
+        evaluator_tensor_product.integrate_gradient<false>(values, gradients);
+        __syncthreads();
+      }
+  }
+
+
+
+  template <int dim, int fe_degree, int n_q_points_1d, int n_components_,
+            typename Number>
+  __device__
+  typename FEEvaluation<dim, fe_degree, n_q_points_1d, n_components_, Number>::value_type
+  FEEvaluation<dim, fe_degree, n_q_points_1d, n_components_, Number>::
+  get_value(const unsigned int q_point) const
+  {
+    return values[q_point];
+  }
+
+
+
+  template <int dim, int fe_degree, int n_q_points_1d, int n_components_,
+            typename Number>
+  __device__ void
+  FEEvaluation<dim, fe_degree, n_q_points_1d, n_components_, Number>::
+  submit_value(const value_type &val_in, const unsigned int q_point)
+  {
+    values[q_point] = val_in * JxW[q_point];
+  }
+
+
+
+  template <int dim, int fe_degree, int n_q_points_1d, int n_components_,
+            typename Number>
+  __device__
+  typename FEEvaluation<dim, fe_degree, n_q_points_1d, n_components_, Number>::gradient_type
+  FEEvaluation<dim, fe_degree, n_q_points_1d, n_components_, Number>::
+  get_gradient(const unsigned int q_point) const
+  {
+    // TODO optimize if the mesh is uniform
+    const Number *inv_jacobian = &inv_jac[q_point];
+    gradient_type grad;
+    for (int d_1=0; d_1<dim; ++d_1)
+      {
+        Number tmp = 0.;
+        for (int d_2=0; d_2<dim; ++d_2)
+          tmp += inv_jacobian[padding_length*n_cells*(dim*d_2+d_1)] *
+                 gradients[d_2][q_point];
+        grad[d_1] = tmp;
+      }
```

Comment:
Right now, the code here (and also in the `read_dof_values` and `distribute_local_to_global` functions above) assumes `n_components_==1`. Can you please add a `static_assert` until the general case is implemented?